src/charm.py

```diff
@@ -1,104 +1,334 @@
+        node_dn = self._get_dn(node)
+        # share node DN with peers
+        if node_dn:
+            self._peers.data[self.unit].update({"dn": node_dn})
+        return node_dn
+
+    def _get_dn(self, cert_path: Path) -> str:
+        """Get the distinguished name from a cert file.
+
+        Args:
+            cert_path (Path): path from cert file.
+
+        Returns:
+            str: DN containing the fields CN, OU, O, L, ST and C
+        """
+        if cert_path.exists():
+            try:
+                logger.info(f" Getting DN from file {cert_path}")
+                dn = (
+                    check_output(f"openssl x509 -subject -noout -in {cert_path}".split())
+                    .decode("ascii")
+                    .rstrip()
+                )
+                return dn.replace("subject=", "").replace(" = ", "=").replace(", ", ",")
+            except CalledProcessError as e:
+                logger.exception(f"Failed to run command {e.cmd}: {e}")
+                raise
+        return None
+
+    @property
+    def _nodes_dn(self) -> List[str]:
+        """Retrieve distinguished names (DN) associated with OpenSearch nodes.
+
+        Returns:
+            List[str]: a list of DN from all nodes in the peer relation.
+        """
+        nodes_dn = [str(self._peers.data[unit].get("dn")) for unit in self._peers.units]
+        self_dn = self._node_dn
+        nodes_dn.append(self_dn)
+        nodes_dn.sort()
+        return nodes_dn
+
+    def _get_context(self) -> Dict[str, Any]:
+        """Get variables context necessary to render configuration templates.
+
+        Returns:
+            Dict[str, Any]: Dictionary containing the variables to be used on
+            jinja2 templates
+        """
+        context = dict(self.model.config)
+        context["unit_ips"] = self._unit_ips
+        context["network_host"] = self._unit_ip
+        context["node_name"] = self._unit_name
+        context["node_cert"] = f"{self._unit_name}.pem"
+        context["node_key"] = f"{self._unit_name}-key.pem"
+        context["admin_dn"] = self._admin_dn
+        context["ca"] = "root-ca.pem"
+        context["nodes_dn"] = self._nodes_dn
+
+        return context
 
     def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+        """Render templates for configuration."""
+        self._configure_security()
+        context = self._get_context()
+        for template in CONFIG_MAP.keys():
+            config_path = CONFIG_MAP[template].get("config_path")
+            cmd = CONFIG_MAP[template].get("cmd")
+            chmod = CONFIG_MAP[template].get("chmod")
+            self._write_config(config_path, template, context, chmod)
+            if cmd:
+                try:
+                    check_call(cmd.split())
+                except CalledProcessError as e:
+                    logger.exception(f"Failed to run command {e.cmd}: {e}")
```

Comment:
Here and elsewhere, cf. the [Juju SDK](https://juju.is/docs/sdk/logging), it is important to not build the logging strings yourself so that the logger can do this for you.